dipy/segment/mrf.pyx

```diff
@@ -0,0 +1,729 @@
+    def prob_neighborhood(self, seg, beta, nclasses):
+        
+        r""" Conditional probability of the label given the neighborhood
+        Equation 2.18 of the Stan Z. Li book.
+
+        Parameters
+        -----------
+        seg : ndarray, 
+                3D tissue segmentation derived from the ICM model
+        beta : float, 
+                scalar that determines the importance of the neighborhood and 
+                the spatial smoothness of the segmentation. 
+                Usually between 0 to 0.5
+        nclasses : int, 
+                    number of tissue classes
+
+        Returns
+        --------
+        PLN : ndarray, 
+                4D probability map of the label given the neighborhood of
+                the voxel
+        """
+
+        cdef:
+            double[:, :, :] P_L_N = np.zeros(seg.shape, dtype=np.float64)
+            cnp.npy_intp classid = 0
+
+        PLN_norm = np.zeros(seg.shape, dtype=np.float64)
+        PLN = np.zeros(seg.shape + (nclasses,), dtype=np.float64)
+
+        for classid in range(nclasses):
+
+            P_L_N = np.zeros(seg.shape, dtype=np.float64)
+            _prob_class_given_neighb(seg, beta, classid, P_L_N)
+
+            PLN[:, :, :, classid] = np.array(P_L_N)
+            PLN[:, :, :, classid] = np.exp(- PLN[:, :, :, classid])
+            PLN_norm += PLN[:, :, :, classid]
+
+        for l in range(nclasses):
+            PLN[:, :, :, l] = PLN[:, :, :, l] / PLN_norm
+
+        return PLN
+
+
+cdef void _initialize_maximum_likelihood(double[:,:,:,:] nloglike, double[:,:,:] seg) nogil:
+    
+    r""" Initializes the segmentation of an image with given neg-log-likelihood
+
+    Initializes the segmentation of an image with neg-log-likelihood field
+    given by `nloglike`. The class of each voxel is selected as the one with
+    the minimum neg-log-likelihood (i.e. the maximum-likelihood segmentation).
+
+    Parameters
+    ----------
+    nloglike : array
+        4D nloglike[x,y,z,k] is the likelihhood of class k for voxel (x,y,z)
+    seg : array
+        3D buffer for the initial segmentation
+        
+    Returns : 
+    seg : array,
+            3D initial segmentation
+    """
+    
+    cdef:
+        cnp.npy_intp nx = nloglike.shape[0]
+        cnp.npy_intp ny = nloglike.shape[1]
+        cnp.npy_intp nz = nloglike.shape[2]
+        cnp.npy_intp nclasses = nloglike.shape[3]
+        double min_energy
+        int best_class
+    for x in range(nx):
+        for y in range(ny):
+            for z in range(nz):
+
+                best_class = -1
+                for k in range(nclasses):
+                    if (best_class == -1) or (nloglike[x,y,z,k] < min_energy):
+                        best_class = k
+                        min_energy = nloglike[x,y,z,k]
+                seg[x,y,z] = best_class
+
+
+cdef void _icm_ising(double[:,:,:,:] nloglike, double beta, double[:,:,:] seg,
+                     double[:,:,:] energy, double[:,:,:] new_seg) nogil:

```

Comment:
Same here, `seg` and `new_seg` don't need to be double, they should be of an integral type